src/traffic_server/traffic_server.cc

```diff
@@ -185,10 +185,21 @@ PID of the traffic_server process (second column of output).
 */
 static int cmd_block = 0;
 
-// 1: the main thread delayed accepting, start accepting.
-// 0: delay accept, wait for cache initialization.
-// -1: cache is already initialized, don't delay.
-static int delay_listen_for_cache = 0;
+namespace
+{
+/** State of waiting for cache to initialize.
+ * Transitions:
+ * WAIT -> DELAYED : wait enabled and proxy initialization was delayed.
+ * WAIT -> INITIALIZED : cache is initialized before delay, don't delay.
+ * DELAYED -> INITIALIZED : cache is initialized, need to proxy init after delay.
+ */
+enum class CacheWaitState {
+  WAIT        = 0, ///< Wait for initialization
+  DELAYED     = 1, ///< The waiting has started.
+  INITIALIZED = -1 ///< Cache is initialized, no more waiting.
+};
+std::atomic<CacheWaitState> delay_listen_for_cache{CacheWaitState::WAIT};
```

Comment:
It's unclear why this needs to be atomic.

Reply:
The cache initialization is proceeding asynchronously and can complete at the same time as the main thread is about to check. Setting to `DELAYED` or `INITIALIZED` can therefore be a race.

Reply:
Because there is a race between setting it to `DELAYED` and `INITIALIZED`.